man/omorfi-analyse.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.40.4.
.TH OMORFI-ANALYSE "1" "March 2015" "OMORFI" "User Commands"
.SH NAME
omorfi-analyse \- Analyse running Finnish text with omorfi python API
.SH SYNOPSIS
.B omorfi-analyse.py
[\fI\-h\fR] [\fI\-f FSAPATH\fR] [\fI\-i INFILE] [\fI-v\fR] [\fI\-o OUTFILE] \-F
FORMAT
.SH DESCRIPTION
Analyses running text using omorfi automata and hfst python library.
.TP
\fB\-h\fR, \fB\-\-help\fR
show this help message and exit
.TP
\fB\-f\fR FSAPATH, \fB\-\-fsa\fR FSAPATH
Path to directory of HFST format automata
.TP
\fB\-i\fR INFILE, \fB\-\-input\fR INFILE
source of analysis data
.TP
\fB\-v\fR, \fB\-\-verbose\fR
print verbosely while processing
.TP
\fB\-f\fR FORMAT, \fB\-\-format\fR FORMAT
output in format compatible with FORMAT
.PP
If no INFILE is given, input is read from standard input. If no OUTFILE is
given, output is written to standard output. Format is required.
.SH EXAMPLES
.TP
Analyse corpus:
omorfi-analyse.py -F vislcg3 \-i rautatie.txt \-o rautatie.vislcg3
.SH COPYRIGHT
Copyright \(co 2015 Omorfi contributors
Licence GPLv3: GNU GPL version 3 <http://gnu.org/licenses/gpl.html>
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


NEWS
# NEWS

This file lists noteworthy changes between releases, for full list of changes,
see git log and then `ChangeLog.old`.

## Significant changes in 20150904

* allomorphy can be tagged again to distinguish e.g. *-iden* and *-itten* when
  generating
* FinnTreeBank-1 format provided by Miikka Silfverberg is available but not
  built by default since it lacks a test set
* lexicalised inflections can have separate tag, e.g. *kännissä* can be lexical
  inessive distinguished from regular inessive
* preliminary VISL CG-3 support, with original grammar by Fred Karlsson;
  convenience bash scripts available for disambiguated parsing
* preliminary support for conllu and conllx analysis formats
* paradigm categorisation is now verified by regular expressions
* lots of paradigm fixes and some added words

## Significant changes in 20150326

* speed is up to >20,000 tokens per second from ~500
* coverages are up to:
  europarl (99 %) gutenberg (97 %), JRC Acquis (94 %) and fiwiki (93 %)
* moses factored model format supported
* segmentation supported
* Java API
* Python hacks packaged to API and module
* Rest of hand-written Xerox legacy data removed; all is script-generated
* github migration since google code is EOL'd
* file naming for automata changed to include omorfi prefix for all file
  names in case they are distributed separately.

## Significant changes in 20141014

* The regressions are also set on coverage over popular corpora:
  Europarl (98 %), FTB 3.1 (97 %), gutenberg (96 %), JRC Acquis (93 %) and
  fiwiki (90 %)
* sti derivation tentatively added
* number of new paradigms and paradigm moves, esp. in old and archaic styles
* some new words manually added
* apertium formats updated totally
* interjection chaining
* rest of hand-written lexc removed: everything in db and python code now
* more strict building and testing altogether (no more dangling references or
  missing tags allowed)
* morphological segmentation should be usable now
* lots of other classifications and attributes added

## Significant changes in 20130829

* Default tag format is now FTB3.1. Recall is 90 % and the format is stable and
  easy to read by humans, which is now the main target for computational
  morphologies.
* The omor tagsets are now permanently unstable and subject to change any day.
  To use them, python scripts have been provided.
* Lots of proper nouns and semantics from Uni Hel projects
* speller build support for new voikko versions
* New regression tests for stuffs
* Most of legacy lexc sources removed; they are now generated from TSV
  "databases".
* The morphological classes now follow 3 main classes with some subclasses that
  are less morphological
* Twol rules and flag diacritics have been eliminated
* Lots of support scripts to verify and extend classifications
* Lots of new word-forms, inflections and changes to derivations
* Some python support scripts for omor formats

## Significant changes in 20121226

* Added fi.wiktionary.org as lexical source (much thanks to students of my unix
  tools course for scripting)
* Added first batch of new proper nouns from a project in Univ. Helsinki
* Lexc data is now rebuild from lexical sources as standard processing;
    * requiring python3
* Minor bug fixes to man pages, special boundaries (e.g. in arkki_tehti)

## Significant changes in 20120401

* Fixed some twol rules w.r.t. new features that blocked compiling
* Autogenerate lexicons from csv data all the time
* Moved to git and googlecode -> chopped most of the documentation and such
* Fixed scripts a bit, added man pages
* Made very crude tests to have at least something back in.

## Significant changes in 20110505

* whole new finntreebank tagset for forthcoming finntreebank work
* uppercasing is noted in the analysis level
* the word boundaries of lexicalised compounds may be available for more cases
  (depending on the tagset)
* whole new lemmatizer tagset is available
* some dozens of new words added and fixed
* combine corpus analysis script with apertium's preprocessors
* causative derivation chain added
* bbreviations, adpositions, prefixes and suffixes are no longer pos but subcat
  analyses


## Significant changes since 20100401

* Include deverbal nouns in compounding system

* Start marking compound and strong morpheme boundaries

* New lexical data handling systems

* Implement generator from analyser

* Subcategorize lots of classes for CG and apertium

* Write documentation in booklet format

* New URI and digit string guessers

* New tagging style colorterm for interactive use

* Include weighting scheme in default build

* Demote SUFFIX from POS reading to SUBCAT

## Significant changes since 20100111

* Added marginal enclitics kA, kAs

* Added LEMMA= structure

* re-organized source code to modules

* Added tagging schemes, weighting schemes and suggestion algorithms

## Significant changes since 0.5

* completely new morphology built on traditional lexc-twolc model

* easier route to add new lexical data via simple CSV format

* lots of new lexical data from Joukahainen project as well as extended
  from kotus-sanalista semi-automatically and *by hand*.

* titlecasing filter for regular words

* š filter for old orthography variants

* compounding much less haphazard concoction

* parts of speech classified and included

* pronouns, interjections, numerals, proper nouns

* much closer to real full fledged morphology

* movement from SFST to HFST toolset with lots of new cool toys (SFST support
  is retained in HFST)

* towards full-scale automatic test suite

<!-- vim: set ft=markdown: -->


src/python/omorfi-analyse.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from sys import stderr, stdin, stdout
from argparse import ArgumentParser, FileType
from omorfi.omorfi import Omorfi

import re

def get_lemmas(anal):
    re_lemma = re.compile("\[WORD_ID=([^]]*)\]")
    lemmas = re_lemma.finditer(anal.output)
    rv = []
    for lemma in lemmas:
        rv += [lemma.group(1)]
    return rv

def print_analyses(surf, anals, format, outfile):
    if format == 'xerox':
        print_analyses_xerox(surf, anals, outfile)
    elif format == 'apertium':
        print_analyses_apertium(surf, anals, outfile)
    elif format == 'vislcg3':
        print_analyses_vislcg3(surf, anals, outfile)
    elif format == 'conllx':
        print_analyses_conllx(surf, anals, outfile)
    elif format == 'conllu':
        print_analyses_conllu(surf, anals, outfile)
    else:
        print("format unknown:", format, file=stderr)
        exit(2)

def print_analyses_xerox(surf, anals, outfile):
    for anal in anals:
        print(surf, anal.output, sep='\t', file=outfile)
    print

def print_analyses_apertium(surf, anals, outfile):
    print("^", surf, sep='')
    for anal in anals:
        print("/", anal.output, sep='', file=outfile)
    print("$")

def print_analyses_vislcg3(surf, anals, outfile):
    print('"<', surf, '>"', sep='', file=outfile)
    re_lemma = re.compile("\[WORD_ID=([^]]*)\]")
    re_pos = re.compile("\[POS=([^]]*)\]")
    re_mrd = re.compile("\[([^=]*)=([^]]*)]")
    for anal in anals:
        pos_matches = re_pos.finditer(anal.output)
        pos = "UNK"
        mrds = []
        lemmas = []
        for pm in pos_matches:
            pos = pm.group(1)
        lemmas = get_lemmas(anal)
        mrd_matches = re_mrd.finditer(anal.output)
        for mm in mrd_matches:
            if mm.group(1) == 'WORD_ID':
                mrds = []
            elif mm.group(1) == 'CASECHANGE' and mm.group(2) != 'NONE':
                mrds = ['<' + mm.group(2) + '>'] + mrds
            elif mm.group(1) == 'ALLO':
                mrds = ['<' + mm.group(2) + '>'] + mrds
            elif mm.group(1) == 'WEIGHT' and mm.group(2) != 'inf':
                    mrds += ['<W=' + str(int(float(mm.group(2)) * 100)) + '>']
            elif mm.group(1) == 'WEIGHT' and mm.group(2) == 'inf':
                    mrds += ['<W=65536>']
            elif mm.group(1) in ['STYLE']:
                mrds += ['<' + mm.group(2) + '>']
            else:
                mrds += [mm.group(2)]
        print('\t"', ''.join(lemmas).replace('"', '\\"'), '" ',
                ' '.join(mrds), sep='', file=outfile)
    print(file=outfile)

def print_analyses_conllx(surf, anals, outfile):
    print(1, surf, "#".join(get_lemmas(anals[0])),
            "-", "-", anals[0].output,
            "-", "-", "-", "-", sep="\t", file=outfile)

def print_analyses_conllu(surf, anals, outfile):
    print(1, surf, "#".join(get_lemmas(anals[0])), 
            "-", "-", anals[0].output,
            "-", "-", "-", "-", sep="\t", file=outfile)

def main():
    """Invoke a simple CLI analyser."""
    a = ArgumentParser()
    a.add_argument('-f', '--fsa', metavar='FSAPATH',
            help="Path to directory of HFST format automata")
    a.add_argument('-i', '--input', metavar="INFILE", type=open,
            dest="infile", help="source of analysis data")
    a.add_argument('-v', '--verbose', action='store_true',
            help="print verbosely while processing")
    a.add_argument('-o', '--output', metavar="OUTFILE", dest="outfile",
            help="print output into OUTFILE", type=FileType('w'))
    a.add_argument('-F', '--format', metavar='FORMAT', required=True,
            help="Output in format compatible with FORMAT",
            choices=['xerox', 'apertium', 'vislcg3', 'conllx', 'conllu'])
    options = a.parse_args()
    omorfi = Omorfi(options.verbose)
    if options.fsa:
        if options.verbose:
            print("reading language models in", options.fsa)
        omorfi.load_from_dir(options.fsa, analyse=True, accept=True)
    else:
        if options.verbose:
            print("reading language models in default dirs")
        omorfi.load_from_dir()
    if not options.infile:
        options.infile = stdin
    if options.verbose:
        print("analysing", options.infile.name)
    if not options.outfile:
        options.outfile = stdout
    if options.verbose:
        print("writing to", options.outfile.name)
    for line in options.infile:
        line = line
        if not line or line == '':
            continue
        surfs = omorfi.tokenise(line)
        for surf in surfs:
            anals = omorfi.analyse(surf)
            print_analyses(surf, anals, options.format, options.outfile)
    exit(0)

if __name__ == "__main__":
    main()
